eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ExternalDependencyExposedCheck.java

```diff
@@ -0,0 +1,218 @@
+                checkNoExternalDependencyExposed(token);
+                break;
+            default:
+                // Checkstyle complains if there's no default block in switch
+                break;
+        }
+    }
+
+    /**
+     * Checks for external dependency, log the error if it is an invalid external dependency.
+     *
+     * @param methodDefToken METHOD_DEF AST node
+     */
+    private void checkNoExternalDependencyExposed(DetailAST methodDefToken) {
+        final DetailAST modifiersToken = methodDefToken.findFirstToken(TokenTypes.MODIFIERS);
+
+        // Getting the modifier of the method to determine if it is 'public' or 'protected'.
+        // Ignore the check if it is neither of 'public' nor 'protected',
+        final AccessModifier accessModifier = CheckUtil.getAccessModifierFromModifiersToken(modifiersToken);
+        if (!accessModifier.equals(AccessModifier.PUBLIC) && !accessModifier.equals(AccessModifier.PROTECTED)) {
+            return;
+        }
+
+        // Checks for the return type of method
+        final DetailAST typeToken = methodDefToken.findFirstToken(TokenTypes.TYPE);
+        if (typeToken != null) {
+            getInvalidReturnTypes(typeToken).forEach(
+                (token, returnTypeName) -> log(token, String.format(EXTERNAL_DEPENDENCY_ERROR, returnTypeName)));
+        }
+
+        // Checks for the parameters of the method
+        final DetailAST parametersToken = methodDefToken.findFirstToken(TokenTypes.PARAMETERS);
+        if (parametersToken != null) {
+            getInvalidParameterTypes(parametersToken).forEach(
+                (token, returnTypeName) -> log(token, String.format(EXTERNAL_DEPENDENCY_ERROR, returnTypeName)));
+        }
+    }
+
+    /**
+     * Get invalid return types from a given TYPE node.
+     *
+     * @param typeToken TYPE AST node
+     * @return a map that maps the invalid TYPE node and the type name.
+     */
+    private Map<DetailAST, String> getInvalidReturnTypes(DetailAST typeToken) {
+        final Map<DetailAST, String> invalidReturnTypeMap = new HashMap<>();
+
+        // Add all invalid external return types to the map
+        final DetailAST identToken = typeToken.findFirstToken(TokenTypes.IDENT);
+        if (identToken == null) {
+            return invalidReturnTypeMap;
+        }
+        final String typeName = identToken.getText();
+        if (!isValidClassDependency(typeName)) {
+            invalidReturnTypeMap.put(typeToken, typeName);
+        }
+
+        // TYPE_ARGUMENTS, add all invalid external types to the map
+        final DetailAST typeArgumentsToken = typeToken.findFirstToken(TokenTypes.TYPE_ARGUMENTS);
+        if (typeArgumentsToken != null) {
+            invalidReturnTypeMap.putAll(getInvalidTypeFromTypeArguments(typeArgumentsToken));
+        }
+
+        // TYPE_ARGUMENT, add the invalid external type to the map
+        final DetailAST typeArgumentToken = typeToken.findFirstToken(TokenTypes.TYPE_ARGUMENT);
```

Comment:
Seems like you're duplicating the TYPE_ARGUMENT check.

What is a method definition where there are **not** TYPE_ARGUMENTS but have a TYPE_ARGUMENT.

Reply:
Example: Even `public Mono<EventDataBatch> createBatch() { }` has TYPE_ARGUMENTS.

![image](https://user-images.githubusercontent.com/10136526/60910379-7ae05780-a235-11e9-90cb-babc4e2cfcd5.png)

Reply:
Two scenarios:
(1)
TYPE_ARGUMENTS
  TYPE_ARGUMENT
  TYPE_ARGUMENT

(2)
TYPE_ARGUMENT

Reply:
Do you have an example where it is possible for scenario 2 to exist?

Reply:
In order to check if the AST node is TYPE_ARGUMENTS or TYPE_ARGUMENT, I had two approaches to handle the condition check. (1) node.findFirstToken() and (2) node.branchContains(). Both have to iterate over a for loop. I choose the current approach, (1) for the simplicity of code. But seems (2) may be easier to understand from their method naming.

Reply:
Ignore the last comment. There is no TYPE_ARGUMENT scenario. Sorry for the confusion.